recognition/include/pcl/recognition/impl/linemod/line_rgbd.hpp

```diff
@@ -150,72 +150,59 @@ LineRGBD<PointXYZT, PointRGBT>::loadTemplates (const std::string &file_name, con
-  for (std::size_t i = 0; i < template_point_clouds_.size (); ++i)
+  for (std::size_t i = 0; i < template_point_clouds_.size (); ++i) 
   {
-    PointCloud<PointXYZRGBA> & template_point_cloud = template_point_clouds_[i];
-    BoundingBoxXYZ & bb = bounding_boxes_[i];
-    bb.x = bb.y = bb.z = std::numeric_limits<float>::max ();
-    bb.width = bb.height = bb.depth = 0.0f;
-
-    float center_x = 0.0f;
-    float center_y = 0.0f;
-    float center_z = 0.0f;
-    float min_x = std::numeric_limits<float>::max ();
-    float min_y = std::numeric_limits<float>::max ();
-    float min_z = std::numeric_limits<float>::max ();
-    float max_x = -std::numeric_limits<float>::max ();
-    float max_y = -std::numeric_limits<float>::max ();
-    float max_z = -std::numeric_limits<float>::max ();
-    std::size_t counter = 0;
-    for (std::size_t j = 0; j < template_point_cloud.size (); ++j)
-    {
-      const PointXYZRGBA & p = template_point_cloud[j];
-
-      if (!std::isfinite (p.x) || !std::isfinite (p.y) || !std::isfinite (p.z))
-        continue;
-
-      min_x = std::min (min_x, p.x);
-      min_y = std::min (min_y, p.y);
-      min_z = std::min (min_z, p.z);
-      max_x = std::max (max_x, p.x);
-      max_y = std::max (max_y, p.y);
-      max_z = std::max (max_z, p.z);
+    bounding_boxes_[i] = computeBoundingBoxAndCenterTemplatePointCloud (template_point_clouds_[i]);
+  }
 
-      center_x += p.x;
-      center_y += p.y;
-      center_z += p.z;
+  return (true);
+}
 
-      ++counter;
-    }
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+template <typename PointXYZT, typename PointRGBT> BoundingBoxXYZ
+pcl::LineRGBD<PointXYZT, PointRGBT>::computeBoundingBoxAndCenterTemplatePointCloud (PointCloud<PointXYZRGBA> & template_point_cloud)
+{
+  BoundingBoxXYZ bb;
+  bb.x = bb.y = bb.z = std::numeric_limits<float>::max ();
+
+  Eigen::Vector4f geometric_center = Eigen::Vector4f::Zero ();
+  Eigen::Vector4f min_pos, max_pos;
+  min_pos.fill (std::numeric_limits<float>::max ());
+  max_pos.fill (std::numeric_limits<float>::lowest ());
+  std::size_t counter = 0;
+  for (const PointXYZRGBA & p : template_point_cloud.points)
+  {
+    if (!isXYZFinite(p))
+      continue;
 
-    center_x /= static_cast<float> (counter);
-    center_y /= static_cast<float> (counter);
-    center_z /= static_cast<float> (counter);
+    min_pos = min_pos.cwiseMin (p.getVector4fMap ());
+    max_pos = max_pos.cwiseMax (p.getVector4fMap ());
 
-    bb.width  = max_x - min_x;
-    bb.height = max_y - min_y;
-    bb.depth  = max_z - min_z;
+    geometric_center += p.getVector4fMap ();
 
-    bb.x = (min_x + bb.width / 2.0f) - center_x - bb.width / 2.0f;
-    bb.y = (min_y + bb.height / 2.0f) - center_y - bb.height / 2.0f;
-    bb.z = (min_z + bb.depth / 2.0f) - center_z - bb.depth / 2.0f;
+    ++counter;
+  }
+  geometric_center /= static_cast<float> (counter);
```

Comment:
Have a look at [this](http://pointclouds.org/documentation/classpcl_1_1_point_cloud.html#a0fb96c26772eca605280d622efe35995) to prevent the need for having a foor loop, **in case the point cloud is dense**. 

dim=4
stride=sizeof(PointXYZRGBA)/sizeof(float) (confirm this is an actual integer even if you did floating point division. it should be 5)
offset=0

Then use eigen function for mean and extracting min and max element.